File: app/voice/personas.py
"""Persona registry for the P29 voice coach.

Three original coaching personas — ATLAS, FORGE, VECTOR — chosen for
contrasting registers (hype / mind-muscle / analytical), not for likeness to
any real or fictional person. See
``docs/prompts/P29_VOICE_COACH_PERSONAS.md`` §0 and §2 for the rationale and
the persona table this module implements.

Voice IDs are deliberately NOT modeled here — a single `voice_id: str` field
can't represent them correctly anyway, since Kokoro and ElevenLabs use
incompatible id schemes per provider. They live in
``data/voice/voice_ids.yaml`` (config, not code — see
``app/voice/voice_ids.py``), keyed by (provider, persona), so swapping voice
or provider is a config edit that never touches this module.
"""

from __future__ import annotations

from typing import Literal, get_args

from pydantic import BaseModel

PersonaKey = Literal["atlas", "forge", "vector"]

PERSONA_KEYS: tuple[PersonaKey, ...] = get_args(PersonaKey)


class Persona(BaseModel):
    """Static metadata for one coaching persona."""

    key: PersonaKey
    name: str
    # Named `tone`, not `register`, to avoid shadowing BaseModel's inherited
    # ABCMeta.register classmethod (pydantic warns on the collision).
    tone: str
    voice_texture: str
    # Shapes the RAG system-prompt tone only (S7) — never alters retrieved
    # facts or safety behaviour (P29 spec §7).
    prompt_fragment: str


PERSONAS: dict[PersonaKey, Persona] = {
    "atlas": Persona(
        key="atlas",
        name="ATLAS",
        tone="Loud, hype, celebratory. Short bursts.",
        voice_texture="deep/booming",
        prompt_fragment=(
            "Speak like ATLAS: loud, hype, celebratory, short bursts. Keep it "
            "to one or two sentences. Never invent facts or soften a safety "
            "warning to sound hype."
        ),
    ),
    "forge": Persona(
        key="forge",
        name="FORGE",
        tone="Slow, gravelly, mind-muscle. Cue-heavy.",
        voice_texture="gravelly/older",
        prompt_fragment=(
            "Speak like FORGE: slow, gravelly, mind-muscle focus, cue-heavy. "
            "Favor concrete technique cues over hype. Never invent facts."
        ),
    ),
    "vector": Persona(
        key="vector",
        name="VECTOR",
        tone="Clipped, deadpan, numeric. Reads state back.",
        voice_texture="flat/neutral",
        prompt_fragment=(
            "Speak like VECTOR: clipped, deadpan, numeric. Read back state "
            "plainly (joint, side, rep number) where relevant. Never invent "
            "facts or numbers not present in the retrieved context."
        ),
    ),
}

File: frontend/src/features/voice/personas.ts
/**
 * Client-side persona config for the P29 voice coach (S4).
 *
 * Mirrors the public shape served by `GET /api/v1/voice/personas`
 * (`app/voice/personas.py` / `app/voice/router.py`) so the UI has a
 * synchronous, zero-latency value to render before that fetch resolves, and
 * something to fall back to if it fails — the persona picker must never be
 * empty on a flaky connection. `useCoachVoice` treats a live fetch (when it
 * adds one) as the source of truth; this is the seed/fallback only.
 *
 * Cue key format is `{persona}.{fault_id}.{severity}` per spec §5 — this
 * module owns the `{persona}` half.
 */

export type PersonaKey = "atlas" | "forge" | "vector"

export const PERSONA_KEYS: readonly PersonaKey[] = ["atlas", "forge", "vector"]

export const DEFAULT_PERSONA: PersonaKey = "atlas"

export interface PersonaMeta {
  readonly key: PersonaKey
  readonly name: string
  readonly tone: string
  readonly voiceTexture: string
}

export const PERSONAS: Readonly<Record<PersonaKey, PersonaMeta>> = {
  atlas: {
    key: "atlas",
    name: "ATLAS",
    tone: "Loud, hype, celebratory. Short bursts.",
    voiceTexture: "deep/booming",
  },
  forge: {
    key: "forge",
    name: "FORGE",
    tone: "Slow, gravelly, mind-muscle. Cue-heavy.",
    voiceTexture: "gravelly/older",
  },
  vector: {
    key: "vector",
    name: "VECTOR",
    tone: "Clipped, deadpan, numeric. Reads state back.",
    voiceTexture: "flat/neutral",
  },
}

/** Narrow an arbitrary string (e.g. a `localStorage` read) to a `PersonaKey`. */
export function isPersonaKey(value: string): value is PersonaKey {
  return (PERSONA_KEYS as readonly string[]).includes(value)
}
